lib/statsd/instrument/datagram_builder.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+class StatsD::Instrument::DatagramBuilder
+  unless Regexp.method_defined?(:match?) # for ruby 2.3
+    module RubyBackports
+      refine Regexp do
+        def match?(str)
+          match(str) != nil
+        end
+      end
+    end
+
+    using RubyBackports
+  end
+
+  def self.unsupported_datagram_types(*types)
+    types.each do |type|
+      define_method(type) do |_, _, _, _|
```

Comment:
Instead of undefining methods in the derived classes, perhaps we should only define those methods in the derived classes.

Reply:
We don't really have a choice to expose all the possible metric types as methods on the `StatsD::Instrument::Client` class, unless we provide different subclasses. I want to make sure we are raising a `NotImplementError`, rather than a `NoMethodError` if somebody uses a metric type that is not supported by their implementation.

We can either use a whitelist approach (your proposal) or a blacklist (current implementation). I chose the blacklist approach because it was less duplication, but I don't feel super strongly about it.

Reply:
We might want to consider client subclasses for extensions that are specific to a backend.  I think it would be completely fine for `NoMethodError` to be raised because a method is called on a client that doesn't support that method.  However, in the case of an extension like `tags` that are datadog specific, I don't see an easy way of extending the method without basically re-implementing it.  So I'm not sure if it is worth doing.